src/Raven.Server/Documents/Queries/Results/TimeSeriesRetriever.cs

```diff
@@ -0,0 +1,815 @@
+
+            if (args[index] == null)
+                return null;
+            
+            if (!(args[index] is Document doc))
+            {
+                if (index == 0 && args[0] is Tuple<Document, Lucene.Net.Documents.Document, IState, Dictionary<string, IndexField>, bool?> tuple)
+                    doc = tuple.Item1;
+                else
+                    return args[index];
+            }
+
+            if (fe.Compound.Count == 1)
+                return doc;
+            
+            return GetFieldFromDocument(fe, doc);
+        }
+
+        private static int GetParameterIndex(DeclaredFunction declaredFunction, StringSegment str)
+        {
+            for (int i = 0; i < declaredFunction.Parameters.Count; i++)
+            {
+                var parameter = declaredFunction.Parameters[i];
+
+                if (str == ((FieldExpression)parameter).FieldValue)
+                    return i;
+            }
+
+            return declaredFunction.Parameters.Count;
+        }
+
+        private static object GetFieldFromDocument(FieldExpression fe, Document document)
+        {
+            if (document == null)
+                return null;
+
+            var currentPart = 1;
+            object val = null;
+            var data = document.Data;
+
+            while (currentPart < fe.Compound.Count)
+            {
+                if (data.TryGetMember(fe.Compound[currentPart], out val) == false)
+                {
+                    var id = currentPart == 1
+                        ? document.Id
+                        : data.TryGetId(out var nestedId) 
+                            ? nestedId : null;
+
+                    throw new InvalidQueryException($"Unable to get the value of '{fe.FieldValueWithoutAlias}' from document '{document.Id}'. " +
+                                                        $"Document '{id}' does not have a property named '{fe.Compound[currentPart]}'.");
+                }
+
+                if (!(val is BlittableJsonReaderObject nested))
+                    break;
+
+                data = nested;
+                currentPart++;
+            }
+
+            return val;
+        }
+
+        private object GetValueFromLoadedTag(FieldExpression fe, SingleResult singleResult)
+        {
+            if (_loadedDocuments == null)
+                _loadedDocuments = new Dictionary<string, Document>();
+
+            var tag = singleResult.Tag?.ToString();
+            if (tag == null)
+                throw new InvalidQueryException("Unable to load document from 'Tag' property of time series entry. 'Tag' is null. " +
+                                                $"Timestamp: '{singleResult.TimeStamp.GetDefaultRavenFormat()}'");
+
+            if (_loadedDocuments.TryGetValue(tag, out var document) == false)
+                _loadedDocuments[tag] = document = _database.DocumentsStorage.Get(_context, tag);
```

Comment:
Do we need to check for null here? Before you put it in the `_loadedDocuments` or return it below

Reply:
right, I missed that.

but now that I'm thinking about it, I'm not sure if we should throw on null document/null tag.
maybe just return null?

It might be useful to support things like this:

```
from x.HeartRate between $start and $end
	load Tag as src
	where src != null and src.Accuracy > 2
```

Reply:
Good point, you are right